backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/utils/VdsCpuUnitPinningHelper.java

```diff
@@ -140,18 +159,37 @@ private List<VdsCpuUnit> allocateDedicatedCpus(List<VdsCpuUnit> cpuTopology, VM
             // coreCount is based on the VM topology
             int coreCount = 0;
             for (int core : getOnlineCores(cpusInChosenSocket)) {
-                List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(cpusInChosenSocket, core);
-                int coreThreads = freeCpusInCore.size();
-                while (coreThreads >= vm.getThreadsPerCpu() &&
-                        cpusToBeAllocated.size() < vm.getNumOfCpus() &&
-                        coreCount / vm.getCpuPerSocket() < amountOfVSocketsInPSockets) {
-                    for (int thread = 0; thread < vm.getThreadsPerCpu() && cpusToBeAllocated.size() < vm.getNumOfCpus(); thread++) {
-                        VdsCpuUnit cpuUnit = freeCpusInCore.remove(0);
-                        cpuUnit.pinVm(vm.getId(), vm.getCpuPinningPolicy());
-                        cpusToBeAllocated.add(cpuUnit);
-                    }
-                    coreCount++;
-                    coreThreads -= vm.getThreadsPerCpu();
+                switch (vm.getCpuPinningPolicy()) {
+                    case DEDICATED:
+                        List<VdsCpuUnit> freeCpusInCore = getFreeCpusInCore(cpusInChosenSocket, core);
+                        int coreThreads = freeCpusInCore.size();
+                        while (coreThreads >= vm.getThreadsPerCpu() &&
+                                cpusToBeAllocated.size() < vm.getNumOfCpus() &&
+                                coreCount / vm.getCpuPerSocket() < amountOfVSocketsInPSockets) {
+                            for (int thread = 0; thread < vm.getThreadsPerCpu() && cpusToBeAllocated.size() < vm.getNumOfCpus(); thread++) {
+                                VdsCpuUnit cpuUnit = freeCpusInCore.remove(0);
+                                cpuUnit.pinVm(vm.getId(), vm.getCpuPinningPolicy());
+                                cpusToBeAllocated.add(cpuUnit);
+                            }
+                            coreCount++;
+                            coreThreads -= vm.getThreadsPerCpu();
+                        }
+                        break;
+                    case ISOLATE_THREADS:
+                        List<VdsCpuUnit> cpusInCore = getCpusInCore(cpusInChosenSocket, core);
+                        if (cpusInCore.stream().anyMatch(VdsCpuUnit::isPinned)) {
+                            continue;
+                        }
+                        if (totalCpuCount < vm.getNumOfCpus() && coreCount / vm.getCpuPerSocket() < amountOfVSocketsInPSockets) {
+                            cpusInCore.forEach(cpu -> {
+                                cpu.pinVm(vm.getId(), vm.getCpuPinningPolicy());
+                                cpusToBeAllocated.add(cpu);
```

Comment:
so if I have a physical core with 2 threads and I run a VM with 1 vCPU there, will the cpuset of the vCPU includes both physical threads? I'm not sure if that's what VDSM expects (or VDSM would "automatically" drop the other threads from the shared pool)
elsewhere, there was a request to offline the other threads (so the vCPU would be pinned to a single physical threads and the other threads in that physical core would be offline to avoid the overhead of the unused thread) - I don't think we would want to do that now, but something to keep in mind
adding @nyoxi

Reply:
Actually in this place it's more about taking the physical resources. But later on I changed the code to cope with multiple pCPUs. For example if the core having pCPUs 0 and 1, and we have the pinning for vCPU 0 for it. The engine will now generate `0#0,1`. When having the domxml back, I saw it as `0#0-1`.

Reply:
Thinking more clearly, on migration VDSM expect 1 pCPU per vCPU..

Reply:
You should include only real vCPUs in the pinning. VDSM will exclude the thread siblings from shard pool based on CPU policy.

For the syntax, it does not really matter. You can either use lists `1,2,3` or ranges `4-6` or any combination of those `1,2,4-6,8`.

Reply:
yeah, I think I added the comment in this place because here I realized that we probably have no indication for pcpus that are being reserved/excluded without pinning

but the last comment (Migration will now have it similar to how I start the VM with vCPU having multiple pCPU.) confused me again - in what scenario would we end up with a vcpu that is exclusively pinned to more than a single pcpu?

Reply:
isolated threads is taking a whole core, with a single thread hardware - you are right, only single pcpu is taken.
when having multiple threads hardware we are taking all of them to one vcpu. we just say that this vcpu can run on each one of them.
example:
if the host is 2 threads, let's say the vm having total 2 vcpus (and no vdsm pin and so on to ease on the example).
vcpu 0 will be pinned to pcpu 0 and 1.
vcpu 1 will be pinned to pcpu 2 and 3.

Reply:
ok, that's what I suspected - if you do this then you don't need the isolate-threads policy, right?
the idea is to keep pinning the vcpu to a specific pcpu but to ensure no *other* vcpu runs on another pcpu that belongs to the same pcore

Reply:
> ok, that's what I suspected - if you do this then you don't need the isolate-threads policy, right? the idea is to keep pinning the vcpu to a specific pcpu but to ensure (at the level of vdsm) that no _other_ vcpu runs on another pcpu that belongs to the same pcore

Then isolate-threads will be same as dedicated but for a core, which is pretty much isolate-threads. why we wouldn't need it?

Why we need the pinning to be to specific pCPU? In that case it's for specific pCore.

Reply:
> Why we need the pinning to be to specific pCPU? In that case it's for specific pCore.

It is a performance feature. Making sure that only specific pCPU is used will avoid potential context switches.